### ISSUE TITLE
Passing context to worker

### ISSUE THREAD

**Issue by kartlee** [RichardKnop/machinery#175]
Hi,

I like to pass the context obtained from rest api request to task submitted for worker. This context has requestId and additional details to track about the original request for the which the task is tied. Is there a way to do it?

**Comment by Rio** [comment]
I'm just running into this myself. We're evaluating machinery for workflows but one of the requirements is context propagation for opentracing and maybe access tokens. If anybody has an already working solution to this we'd all like to know. Before we start doing double work ;)

**Comment by Rio** [comment]
I've just noticed `signature.Headers`, might be an idea to add it to the context if not empty? Task should get a reference to the signatures then.

**Comment by Rio** [comment]
Another way would be to extend `Signature` with `Context` that is populated by the various `Server.Send*` methods that propagate the contexts in the case of `SendChain`.

Sorry, just brain dumping.

**Comment by RichardKnop** [comment]
@Rio Sorry for late reply but I have been quite busy at my new job so didn't have much time to spend on machinery.

There is a partial support for contexts already. Take a look at this boolean flag:

https://github.com/RichardKnop/machinery/blob/64d4c50bbd0e943e823596645c04ce98ca1c1830/v1/tasks/task.go#L21

Therefor you might be able to use contexts already. You would need to define a task function with first argument being a context object and then in the signature have first argument of type `context`.

This should then be propagated and a new `context.Background()` object should be passed to the task when it will be executed.

https://github.com/RichardKnop/machinery/blob/64d4c50bbd0e943e823596645c04ce98ca1c1830/v1/tasks/task.go#L72

This was a PR / contribution by somebody else so I am not 100% certain how it works tbh. Give it a try and let me know if it's sufficient.

If not, perhaps you can refactor / change this functionality to better support your use case.

**Comment by Rio** [comment]
I didn't expect any reply soon so that's ok. We're all busy ;)

I've seen the `UseContext` field and was thinking of loading the contents in `Signature.Headers` into the new context that's passed to the function. This way the payload structure doesn't need to be changed unless you foresee some issues with AMPQ with which I'm not familiar.

We could also scope the context within the headers with it's own key and type. This way we can ensure that what ever's put into the context is json serializable.

Then when sending the tasks we could have `SendTaskWithContext` equivalents for all those functions.

I'll have a go at it and we'll see where it ends up.

**Comment by Rio** [comment]
Oh yeah, if we could add a `context` type argument to the signatures that would also work.
Missed your remark about that.

**Comment by seonixx** [comment]
Any updates on this? Can't seem to figure out how to pass the context in to the task signature.

**Comment by Rio** [comment]
I've actually worked on this last week. It's functional but only for passing opentracing contexts. All other values are currently ignored. I'll make a PR but I'd like to put in some tests, haven't come around to that yet and the readme needs work.

**Comment by Rio** [comment]
PR is here: https://github.com/RichardKnop/machinery/pull/248

**Comment by cherubmiki** [comment]
This PR helped us too. However, it would be great if we could set the tracer to the server not just use the global tracer.

What do you think about it?

**Comment by Rio** [comment]
I think having your loggers, metrics and tracers as a dependency is generally a good idea. I think this should be added to the `config.Config` struct that defaults to GlobalTracer if not set.

**Comment by seonixx** [comment]
@RichardKnop regarding the above "partial support" for contexts, unless I'm missing something it would appear that this simply attaches a background context to a task function definition if the first argument in that definition is of type context, however I don't see anyway you can pass an existing context to the task itself as context.Context is not a supported base type for an argument.

@Rio your PR for send with context applies specifically to opentracing is that correct? Not sending of the serialised context as a whole to the worker? From what I can see this is attaching the span to the header, but not passing any additional context along.

In short, I'm trying to figure out, as the OP asked, how to send the context to the worker as an argument. Not just the opentracing component.

I'm fairly new to opentracing so if I'm missing something, my apologies, but I'm needing to pass context to the worker for visibility of session information. Maybe this can/should be done through opentrace? Any input appreciated!

**Comment by RichardKnop** [comment]
@seonixx Do you want to pass a specific context to a task at a point of sending it? This is difficult because machinery is distributed system so tasks are run by different processes / servers.

We could add support for context arguments but I am not 100% sure if contexts are JSON serialisable in a useful way (to at least keep the data associated with them). I would need to experiment with this.

Another way would be to allow passing specific context from worker instance (perhaps define a function to construct context for a task) but again this would be done on a remote server or in different process so not sure it's that useful.

I'm open to PRs to improve this area if people see some straightforward ways to allow passing contexts in a useful way to tasks running on remote servers. Testing marshaling/unmarshaling contexts as JSON might be a good first step.

**Comment by Rio** [comment]
yes the main problem would be that any datatype can live inside a context and we have no way of quickly knowing if they can be serialized to json and if they can be it might not make sense when they get deserialized back on the worker side. This is the reason that contexts for now are opentracing specific as it defines a standard way of serializing/deserializing.

A solution could be that we have a machinery specific type that we know how to serialize and deserialize just like the span context so we can guarantee that it will reach the workers.
  
You could, for now, use the baggage feature of opentracing to get data across. How ever that baggage will travel with the trace for as long as the trace is alive, and depending on the size it could add a significant network cost. Also binding your business logic to opentracing is generally not something that you'd want to do.

**Comment by Rio** [comment]
For example the go libraries for gRPC use a MD or metadata structure that they know how to serialize and deserialize when sending it over the network. Maybe we could have a `MachineryMD`?

**Comment by seonixx** [comment]
Thanks for your input! @RichardKnop you mentioned sending context is difficult because of machinery being distributed, but if the context is serialised and sent as a task argument then is this still an issue?? Just trying to work out where this becomes a problem.

@Rio yes I also looked at the baggage option (after looking into opentracing in more detail) but thought the overhead would become a potential issue at some point.

I'm using go-kit (https://github.com/go-kit) for microservices over HTTP-JSON which provides context threading across services via built-in instrumentation. I might try to find out more about how this works and see if could/should be applied to machinery in a similar way.

@RichardKnop @Rio cheers for your help so far :)

**Comment by jackielii** [comment]
I think injecting the signature into the context is quite useful see PR #401

**Comment by jackielii** [comment]
I think we can close this issue. as the signature is now injected into the task function: https://github.com/jackielii/learn-machinery

The downside is we the header becomes `map[string]interface{}`

maybe we should use json.RawMessage and expose a function called `DecodeHeader` on the signature
